### PR TITLE
[loader-v2] More tests for global manager

### PR DESCRIPTION
**Pull request by georgemitenkov**
## Description

Addressing loader V2 todo: adding more test cases for module manager cache flushing conditions.

## How Has This Been Tested?

N/A

## Key Areas to Review

Ensure all corner cases are covered.

## Type of Change

- [x] Tests

## Which Components or Systems Does This Change Impact?

- [x] Move/Aptos Virtual Machine

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by georgemitenkov** [comment]
* **#15563** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/15563?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/15563?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: aptos-move/block-executor/src/code_cache_global_manager.rs [hunk 1]
```diff
@@ -368,28 +402,185 @@ mod test {
...
+                }
+            });
+            handles.push(handle);
+        }
 
-    // TODO(loader_v2): Add more unit tests like with previous commits.
+        let mut num_guards = 0;
+        let mut num_nones = 0;
+        for handle in handles {
+            if handle.join().unwrap() {
```
Comment (gelash) [inline aptos-move/block-executor/src/code_cache_global_manager.rs:576]: hm, one of them can call try_lock after you've joined the one that previously succeeded, right? so the test can flake? If I'm not missing anything, then one quick fix would be to add an atomic and increment after try lock, and only join when they all increment it.

**Review by gelash — APPROVED** [review]

File: aptos-move/block-executor/src/code_cache_global_manager.rs [hunk 2]
```diff
@@ -368,28 +402,185 @@ mod test {
...
+                num_guards += 1;
+            } else {
+                num_nones += 1;
+            }
+        }
+
+        assert!(num_guards <= 1);
```
Comment (gelash) [inline aptos-move/block-executor/src/code_cache_global_manager.rs:583]: why not assert_eq?

File: aptos-move/block-executor/src/code_cache_global_manager.rs (hunk 1 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/code_cache_global_manager.rs:576]: Indeed.... good catch

File: aptos-move/block-executor/src/code_cache_global_manager.rs (hunk 2 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/code_cache_global_manager.rs:583]: Should be =, yes!

File: aptos-move/block-executor/src/code_cache_global_manager.rs (hunk 1 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/code_cache_global_manager.rs:576]: I think it will always be flaky? the scheduling can always do: thread 1 fully completes, then thread 2 fully completes. I added sleeping to make sure those run concurrently in practice - should be ok?

**Review by zekun000 — APPROVED** [review]

File: aptos-move/block-executor/src/code_cache_global_manager.rs (hunk 1 above)
Comment (gelash) [reply aptos-move/block-executor/src/code_cache_global_manager.rs]: I meant you can share more state to introduce extra barrier after try_lock and before they return, say increment a counter and keep loading until the value is 8 (so they all called try_lock already), then they can clear it and return but the interleavings is restricted, and tests try_lock from different threads.

**Review by runtian-zhou — APPROVED** [review]

File: aptos-move/block-executor/src/code_cache_global_manager.rs (hunk 1 above)
Comment (georgemitenkov) [reply aptos-move/block-executor/src/code_cache_global_manager.rs]: done

**Review by gelash — APPROVED** [review]
